### PR TITLE
Cabal: Add flag to ignore build tool dependencies

### LINKED ISSUE #10061: build-tool-depends in Cabal regresses ghc ability to build from source distribution
It should be possible to build ghc from its source distribution with only `ghc`.

When the source distribution is created the .hs file is generated from the .x file. This removes the need for alex when building the project.

Recent changes in Cabal regress this by introducing `build-tool-depends: alex`.

https://gitlab.haskell.org/ghc/ghc/-/issues/24826

Potential solution: Add a flag to `Cabal-syntax` which allows the usage of `build-tool-depends` to be disabled. We can then enable this in hadrian (which ignores build-tool-depends).

One workaround we can do in hadrian is to always pass the `--with-alex` and `--with-happy` flags when configuring packages using `Cabal` because `hadrian` deals with processing these files. (Perhaps there are other workarounds?)

### PR DESCRIPTION
**Pull request by alt-romes**
Add a flag to disable the hard requirement on the
build-tools-(depends) declared on the Cabal package.

When this flag is enabled (--ignore-build-tools), a build-tool which can't be found does not block compilation.

Fixes #10061

Please read [Github PR Conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#github-pull-request-conventions) and then fill in *one* of these two templates.

---

**Template Α: This PR modifies [behaviour or interface](https://github.com/cabalism/cabal/blob/master/CONTRIBUTING.md#changelog)**

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.
* [x] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [x] Tests have been added. (*Ask for help if you don’t know how to write them! Ask for an exemption if tests are too complex for too little coverage!*)

### CODE REVIEW

**Review by ulysses4ever — APPROVED** [review]
Thanks! 

Adding a flag is a user-visible change -- please, add a changelog file. I'm not sure why the corresponding checkbox in the PR description is ticked.

**Comment by alt-romes** [comment]
> Thanks!
> 
> Adding a flag is a user-visible change -- please, add a changelog file. I'm not sure why the corresponding checkbox in the PR description is ticked.

I thought that this kind of flag is not really important to announce, as its use case is quite niche. I can if you still think it is relevant to the user in general.

**Comment by ulysses4ever** [comment]
I think we ought to document the addition of any flag just like with any user-visible change, however subtle the use case for the flag is.

**Comment by alt-romes** [comment]
> I think we ought to document the addition of any flag just like with any user-visible change, however subtle the use case for the flag is.

I've added the changelog

File: Cabal/src/Distribution/Simple/Configure.hs [hunk 1]
```diff
@@ -848,29 +848,33 @@ configurePackage cfg lbc0 pkg_descr00 flags enabled comp platform programDb0 pac
...
-        let unknownBuildTools =
-              [ buildTool
-              | buildTool <- buildTools bi
-              , Nothing == desugarBuildTool pkg_descr0 buildTool
-              ]
-        externBuildToolDeps ++ unknownBuildTools
+  let requiredBuildTools
+        -- If --ignore-build-tools is set, no build tool is required:
+        | fromFlagOrDefault False $ configIgnoreBuildTools cfg =
+            []
```
Comment (andreabedini) [inline Cabal/src/Distribution/Simple/Configure.hs:854]: Could we instead leave this unchanged and skip the `configureRequiredPrograms verbosity requiredBuildTools` that is after this?
The build tools are still required but we just skip checking for them.

**Review by andreabedini — COMMENTED** [review]
I am not sure whether "ignore" is the right concept here.

In my mind I was thinking we could skip checking for them ahead of time.

Of course if a build-tool is actually required, Cabal will have to run it and it will fail is the tool is not available.

@ulysses4ever @Kleidukos @Mikolaj what do you think?

PS: I do not believe that this change is necessary but I still think it can be useful in some situations (e.g. the tool is not available during configuration but will be available during build).

File: Cabal/src/Distribution/Simple/Setup/Config.hs [hunk 2]
```diff
@@ -319,7 +323,9 @@ instance Eq ConfigFlags where
       && equal configDebugInfo
       && equal configDumpBuildInfo
       && equal configUseResponseFiles
+      && equal configAllowDependingOnPrivateLibs
```
Comment (geekosaur) [inline Cabal/src/Distribution/Simple/Setup/Config.hs:326]: Is this a leak from the private dependencies PR?

**Review by geekosaur — APPROVED** [review]
Otherwise looks okay, but I can't help but think this is the wrong way to handle the problem it's supposed to work around.

File: Cabal/src/Distribution/Simple/Setup/Config.hs (hunk 2 above)
Comment (alt-romes) [reply Cabal/src/Distribution/Simple/Setup/Config.hs:326]: No, this was missing from the `Eq` check. It's about private sub-libraries (as in, not visible)

**Comment by ulysses4ever** [comment]
@andreabedini I think the current design is "good enough": it sounds reasonably general (you can imagine circumstances where you want to ignore the field), and, on the other hand, it will help a particular Cabal user, GHC (a rather "big" user too).

I can imagine reasonable people may disagree on technical details of implementation or even the design itself (e.g. whether to call it "ignore"). I don't have anything to contribute here, I'm afraid. I'd like us to converge on a solution that on one hand helps GHC achieve its goals, and on the other, doesn't complicate the Cabal codebase or mental model too much. The patch seems tiny, at least,  so for one, it doesn't seem to bring in much complexity.

This was discussed at today's Cabal meeting at length, and everyone agreed that we spend a lot of energy on a rather tiny thing. If we (the Cabal side) want to be extra conservative and reject this change, GHC surely can survive and work around it. But then no one sees a good reason to be extra conservative on this particular issue, as well as to split any more hair on it. Plus, the implementation is already here.

File: Cabal/src/Distribution/Simple/Setup/Config.hs (hunk 2 above)
Comment (ulysses4ever) [reply Cabal/src/Distribution/Simple/Setup/Config.hs:326]: Good catch!

File: Cabal/src/Distribution/Simple/Configure.hs (hunk 1 above)
Comment (alt-romes) [reply Cabal/src/Distribution/Simple/Configure.hs:854]: Sure. What do you suggest the flag should be called ?

Comment (alt-romes) [reply Cabal/src/Distribution/Simple/Configure.hs:854]: Perhaps `--skip-build-tools-check`?

**Comment by andreabedini** [comment]
@ulysses4ever I am ok with this implementation, I am no intention to bikeshed the flag name or other details. IMHO the difference with #10104 is not a mere technical detail but I am aware that this is not the right place to for this discussion.

**Comment by alt-romes** [comment]
> I am ok with this implementation, I am no intention to bikeshed the flag name or other details.

Thanks Andrea. I'm proceeding with assigning this to be merged.

**Comment by ulysses4ever** [comment]
@geekosaur as a manager for the 3.12 branch, do you want to backport this so that it appears in 3.12.2? I think it'd make sense. I haven't checked if this is allowed by PVP though. Please, do, if you want to backport.

**Comment by geekosaur** [comment]
PVP isn't really relevant here because it's a program's command line arguments, not a library API. Working by analogy, though, it adds a new command-line "API" while maintaining backward compatibility,  so can be a minor version bump.

**Comment by geekosaur** [comment]
@mergify backport 3.12

**Comment by geekosaur** [comment]
Welp. Options are versioned and the new option is versioned under 3.13.0. I think that means no-go. I'm closing the backport.

**Comment by alt-romes** [comment]
> Welp. Options are versioned and the new option is versioned under 3.13.0. I think that means no-go. I'm closing the backport.

I don't think that's quite right, or maybe I'm not getting it. `cabal-install` filters out flags that cabal-install will pass to Cabal depending on the Cabal the library version it is using under the hood. This is just a flag in Cabal the library. If you backport it to Cabal 3.12.2, that means Cabal 3.12.2 will support that flag as well. Cabal-install knows nothing of this.

**Comment by geekosaur** [comment]
The problem is that it won't be passed unless the Cabal version is at least 3.13.0.

**Comment by alt-romes** [comment]
> The problem is that it won't be passed unless the Cabal version is at least 3.13.0.

cabal-install will *never* pass this flag, because it doesn't know about it (unless it is specified as an "opaque" flag to pass to Cabal invocations).

This flag is meant for low-level uses of Cabal the library, e.g. Hadrian will use it.

**Comment by geekosaur** [comment]
https://github.com/haskell/cabal/pull/10159/files#diff-
Does this not need to be changed to `flags_3_12_2_0` on both `master` and `3.12` if we backport?

**Comment by alt-romes** [comment]
> https://github.com/haskell/cabal/pull/10159/files#diff-e3eb6f2f1835c361389335e82b4563e49902174f1a3b7f4da50a6627e75783f6 Does this not need to be changed to `flags_3_12_2_0` on both `master` and `3.12` if we backport?

Yes.. 🤦. I thought I hadn't added this logic, but I'm just blind (didn't spot it when looking at the diff just a few moments ago). I guess we could change that too given the conflict... I'm not sure how pressing it is (@mpickering will know). Sorry Brandon!

**Pull request merged**